### PR TITLE
also publish dev version on play store (aka nightlies)

### PR DESCRIPTION
**Pull request by 3nids**

### CODE REVIEW

File: version.pri [hunk 1]
```diff
@@ -1,8 +1,9 @@
 VERSION_MAJOR = 1
 VERSION_MINOR = 4
 VERSION_FIX = 0
-VERSION_RC = 99
-VERSION_SUFFIX = ''
+# for RC (v1.2.3-rc4) or dev (v1.2.3-dev456)
+VERSION_NUMBER = 999
```
Comment (3nids) [inline version.pri]: this has been increased to 3 digits, in case we have more than 99 commits on the dev (nightlies) build.
I guess it won't cause any trouble to the current build, since the global number will have 1 more digit (hence bigger).
Or is there a maximum value for this number (and I would be above now)
@mbernasocchi can you confirm?

File: scripts/docker-build.sh [hunk 2]
```diff
@@ -44,17 +44,20 @@ if [[ "X${PKG_NAME}" != "Xqfield" ]]; then
...
 
-  export RC_SUFFIX=$(echo "${VERSION}" | cut -f 2 -d 'c' -s)
-  sed -i "s/^VERSION_RC\s*= .*/VERSION_RC = ${RC_SUFFIX:-99}/g" ${SOURCE_DIR}/version.pri
-  sed -i "s/^VERSION_SUFFIX\s*= .*/VERSION_SUFFIX = '${RC_SUFFIX:+-rc}$RC_SUFFIX'/g" ${SOURCE_DIR}/version.pri
+  # for RC or dev releases, append the version number (up to 3 digits)
+  export VERSION_NUMBER=$(echo "${APP_VERSION}" | sed -r -e 's/.*?-(\w+)(\d+)/\2/;t;d') # v.1.2.3-dev456 => 456, v.1.2.3 => EMPTY
+  sed -i "s/^VERSION_NUMBER\s*= .*/VERSION_NUMBER = ${VERSION_NUMBER:999}/" ${SOURCE_DIR}/version.pri
+
+  # coming fron env var, might be empty
+  sed -i "s/^VERSION_SUFFIX_STR\s*= .*/VERSION_SUFFIX_STR = ${VERSION_SUFFIX_STR}/" ${SOURCE_DIR}/version.pri
```
Comment (3nids) [inline scripts/docker-build.sh]: instead of being calculated here, it's defined ci/travis-build.sh:
https://github.com/opengisch/QField/pull/781/files#diff-e438ed61bbdfb465bfa5957775fa6912R20
and
https://github.com/opengisch/QField/pull/781/files#diff-e438ed61bbdfb465bfa5957775fa6912R27
so we have more flexibility (commit number added to version string)

File: scripts/deploy.sh [hunk 3]
```diff
@@ -3,30 +3,52 @@ set -e
 
 source scripts/travis_env.sh
 
+ARCHS=( armv7 x86 arm64_v8a x86_64 )
```
Comment (3nids) [inline scripts/deploy.sh]: the main changes here are this array being added to avoid rewriting everywhere the same thing

File: scripts/deploy.sh [hunk 4]
```diff
@@ -3,30 +3,52 @@ set -e
...
+    ASSETS=""
+    for ARCH in "${ARCHS[@]}"
+    do
+      echo -e "\e[93m * Collecting apks to upload...\e[0m"
+      curl -L -s -S -o /tmp/qfield-dev-${ARCH}.apk https://download.opengis.ch/qfield/ci-builds/qfield-dev-${UPLOAD_ARTIFACT_ID}-${TRAVIS_COMMIT}-${ARCH}.apk
+      ASSETS="${ASSETS} /tmp/qfield-dev-${ARCH}.apk"
+    done
+
+    echo -e "\e[93m * Deploying app to google play (release version)...\e[0m"
+    ./scripts/basic_upload_apks_service_account.py ch.opengis.qfield ${ASSETS}
```
Comment (3nids) [inline scripts/deploy.sh]: here is the deploy to play store of the dev version

File: version.pri (hunk 1 above)
Comment (mbernasocchi) [reply version.pri]: the main issue is that ve are getting one order of magnitude closer to the max version size.
https://stackoverflow.com/questions/24246143/maximum-length-of-android-versionname-versioncode-manifest

**Comment by 3nids** [comment]
> not really happy about adding one more 9 to the version number.

@mbernasocchi  OK, but I don't see many alternatives:

* switch to dates: `YYYYMMDDIA` (`I` an increment to allow up to 10 builds per day, `A` the architecture) but that bring us very close to the upper bound limit, no way to add one more digit
* totally rely on commit count to set the version number: this would be an easy way and will always increment. We just need to add a constant to be above current version. eg: `20000000 + commit_count`
* keep track of our own counter (with the same offset than previous solution)
* or the proposed 3 digits in the current PR, but only applied for the nightly builds

(current version is `10200995`)

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-781-1f7426d1379ecfb93c770e5251e63abdcb291277-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-781-1f7426d1379ecfb93c770e5251e63abdcb291277-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-781-1f7426d1379ecfb93c770e5251e63abdcb291277-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-781-1f7426d1379ecfb93c770e5251e63abdcb291277-x86_64.apk)

**Comment by mbernasocchi** [comment]
>     2\. totally rely on commit count to set the version number: this would be an easy way and will always increment. We just need to add a constant to be above current version. eg: `20000000 + commit_count`

why don't we just do the commit number? we can kill the current app id and start from 0 with a new id anyway. the package number is completely indipendent from the release one

**Comment by 3nids** [comment]
@mbernasocchi indeed, good point

File: .travis.yml [hunk 5]
```diff
@@ -17,13 +17,9 @@ before_install:
   - openssl aes-256-cbc -K $encrypted_c158cd588669_key -iv $encrypted_c158cd588669_iv -in keystore.p12.enc -out keystore.p12 -d
 
 script:
-  - export QFIELD_SDK_VERSION=$(awk -F "=" '/osgeo4a_version/{print $2}' sdk.conf)
-  - echo "travis_fold:start:docker-pull"
-  - docker pull opengisch/qfield-sdk:${QFIELD_SDK_VERSION}
-  - echo "travis_fold:end:docker-pull"
-  - echo "travis_fold:start:tx-pull"
+  - export QFIELD_SDK_VERSION=$(cat sdk.conf | cut -d= -f2)
```
Comment (m-kuhn) [inline .travis.yml]: Why did you change this line?
The idea is to add versions for other platforms too, and we won't be able to rely on only one entry being in there.

File: scripts/docker-build.sh [hunk 6]
```diff
@@ -71,14 +74,14 @@ if [ -n "${KEYNAME}" ]; then
 	    --sign ${SOURCE_DIR}/keystore.p12 "${KEYNAME}" \
 	    --storepass "${STOREPASS}" \
 	    --keypass "${KEYPASS}" \
-        --input ${BUILD_DIR}/src/app/android-libqfield.so-deployment-settings.json \
+      --input ${BUILD_DIR}/src/app/android-libqfield.so-deployment-settings.json \
```
Comment (m-kuhn) [inline scripts/docker-build.sh]: dedent?

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-781-c57de4786427f296cd81d79842a3f3417b2d6c63-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-781-c57de4786427f296cd81d79842a3f3417b2d6c63-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-781-c57de4786427f296cd81d79842a3f3417b2d6c63-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-781-c57de4786427f296cd81d79842a3f3417b2d6c63-x86_64.apk)

File: .travis.yml (hunk 5 above)
Comment (3nids) [reply .travis.yml]: Nothing has changed to me here. Still looking in sdk.conf

**Comment by 3nids** [comment]
The main question is that I have removed the logic from version.pri and moved it to docker-build.sh
The logic was spread over Travis-build, docker-build and the version.pri, now it is a bit more centralized and more obvious I hope.
Does it work for you? 
Is it future proof for windows ? Or we don't care for now?

File: .travis.yml (hunk 5 above)
Comment (m-kuhn) [reply .travis.yml]: But on the first line instead of the line for osgeo4a?

**Comment by m-kuhn** [comment]
One thing I'm not sure is how it works for local builds. Setting the arch version is now in travis yml, so if we do a local build for a customer will there be an extra step manually?
Apart from this I'm not particularly attached to the old way of doing things, so if it's more centralized and still works, even better.

File: .travis.yml (hunk 5 above)
Comment (3nids) [reply .travis.yml]: Will revert

**Comment by 3nids** [comment]
It's all in docker-build, so pretty flexible and would work as is

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-781-72756bd1d1c9a3fed028552279010da1ccae8190-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-781-72756bd1d1c9a3fed028552279010da1ccae8190-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-781-72756bd1d1c9a3fed028552279010da1ccae8190-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-781-72756bd1d1c9a3fed028552279010da1ccae8190-x86_64.apk)

**Comment by 3nids** [comment]
ok, now the method to determine the version code has been centralized here (and it's tested): https://github.com/opengisch/QField/pull/781/files#diff-d88da2f0ae98e992fab233374e8f3cb7

I also added a check to avoid creating too big version numbers
https://github.com/opengisch/QField/pull/781/files#diff-268efe2a5282875548af904d51a8577dR62
we can slowly increase it when needed

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-781-939c8d47ae2d57364b03659301473199445194f6-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-781-939c8d47ae2d57364b03659301473199445194f6-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-781-939c8d47ae2d57364b03659301473199445194f6-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-781-939c8d47ae2d57364b03659301473199445194f6-x86_64.apk)

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-781-86aeb881a10c974a94842470926e2d1f4ce6ee74-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-781-86aeb881a10c974a94842470926e2d1f4ce6ee74-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-781-86aeb881a10c974a94842470926e2d1f4ce6ee74-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-781-86aeb881a10c974a94842470926e2d1f4ce6ee74-x86_64.apk)

File: scripts/docker-build.sh [hunk 7]
```diff
@@ -19,22 +26,17 @@ if [[ -z ${BUILD_FOLDER+x} ]]; then
...
-    ARCH=armv7
-fi
-if [[ -z ${APP_NAME+x} ]]; then
-  APP_NAME="QField"
-fi
-if [[ -z ${PKG_NAME+x} ]]; then
-  PKG_NAME="qfield"
-fi
+
+[[ -z ${ARCH} ]] && ARCH=armv7
```
Comment (m-kuhn) [inline scripts/docker-build.sh]: What was the  reason to change this?

File: .travis.yml [hunk 8]
```diff
@@ -37,36 +35,39 @@ jobs:
         - echo "travis_fold:start:build\n$(tput bold)Build QField $(tput sgr0)"
         - ./scripts/ci/pull_translations.sh
         - docker-compose -f .docker/testing/docker-compose-travis.yml run qgis /usr/src/.docker/testing/build-test.sh
-      if: type = pull_request OR tag IS present OR branch = master
+
+    - stage: test
```
Comment (m-kuhn) [inline .travis.yml]: We are already now quite often short on parallel travis jobs, is it required to add one more?

File: scripts/docker-build.sh [hunk 9]
```diff
@@ -44,17 +46,22 @@ if [[ "X${PKG_NAME}" != "Xqfield" ]]; then
...
-  echo "Building release version ${VERSION}"
-  sed -i "s/^VERSION_MAJOR\s*= .*/VERSION_MAJOR = $(echo "${VERSION}" | cut -f 2 -d 'v' | cut -f 1 -d '.')/g" ${SOURCE_DIR}/version.pri
-  sed -i "s/^VERSION_MINOR\s*= .*/VERSION_MINOR = $(echo "${VERSION}" | cut -f 2 -d '.')/g" ${SOURCE_DIR}/version.pri
-  sed -i "s/^VERSION_FIX\s*= .*/VERSION_FIX = $(echo "${VERSION}" | cut -f 3 -d '.' | cut -f 1 -d '-')/g" ${SOURCE_DIR}/version.pri
+  echo "Building release version ${APP_VERSION}"
+  APP_VERSION_CODE=$(app_version_code "${APP_VERSION}")
+  echo "Generated version code: ${APP_VERSION_CODE}"
+fi
+
+# coming fron env var
```
Comment (m-kuhn) [inline scripts/docker-build.sh]:
```suggestion
# coming from env var
```

File: scripts/version_number.sh [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+  elif ([[ $ARCH == x86_64 ]]); then
+    echo 4
+  else
+    echo 0
+  fi
+}
+
+# version to version code
+# arg1: APP_VERSION (e.g. v1.2.3 or v1.2.3-rc4)
+# arg1: ARCH
```
Comment (m-kuhn) [inline scripts/version_number.sh]:
```suggestion
# arg2: ARCH
```

File: scripts/ci/travis_build.sh [hunk 11]
```diff
@@ -1,19 +1,56 @@
...
 if [ -n "${TRAVIS_TAG}" ]; then
   echo "Building release"
   export APP_NAME="QField"
   export PKG_NAME="qfield"
   export APP_ICON="qfield-logo.svg"
+  export APP_VERSION="${TRAVIS_TAG}" #  (v1.2.3 or v1.2.3-rc4)
+
+  # adding the code name to the version string
+  CURRENT_CODENAME=$(cat ${DIR}/../../RELEASE_NAME)
+  APP_VERSION_STR="${TRAVIS_TAG} - ${CURRENT_CODENAME}"
```
Comment (m-kuhn) [inline scripts/ci/travis_build.sh]: The initial `v` needs to be stripped from TRAVIS_TAG

File: RELEASE_NAME [hunk 12]
```diff
@@ -0,0 +1 @@
+Olavtoppen
```
Comment (m-kuhn) [inline RELEASE_NAME]: This needs to be updated before every release, correct?

File: scripts/docker-build.sh (hunk 7 above)
Comment (m-kuhn) [inline scripts/docker-build.sh]:
```suggestion
# Set default values if missing
[[ -z ${ARCH} ]] && ARCH=armv7
```

File: .travis.yml (hunk 8 above)
Comment (3nids) [reply .travis.yml]: ok, I can drop it, although it lasts 40 sec

File: RELEASE_NAME (hunk 12 above)
Comment (3nids) [reply RELEASE_NAME]: yes, we could make it a bit more clever by looking in the release message, but it's quite some code and complexity. I don't think it's worth it.

File: .travis.yml (hunk 8 above)
Comment (3nids) [reply .travis.yml]: done

File: scripts/ci/travis_build.sh (hunk 11 above)
Comment (3nids) [reply scripts/ci/travis_build.sh]: fixed, use a dedicated and tested function

File: RELEASE_NAME (hunk 12 above)
Comment (m-kuhn) [reply RELEASE_NAME]: Agreed, I thought the same, it was only to be sure :+1:

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-781-421118ee53125e21009496abe8f8fffaaf454c80-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-781-421118ee53125e21009496abe8f8fffaaf454c80-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-781-421118ee53125e21009496abe8f8fffaaf454c80-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-781-421118ee53125e21009496abe8f8fffaaf454c80-x86_64.apk)

File: scripts/deploy.sh [hunk 13]
```diff
@@ -59,6 +59,6 @@ if [[ ${TRAVIS_SECURE_ENV_VARS} = true ]]; then
     done
 
     echo -e "\e[93m * Deploying app to google play (release version)...\e[0m"
-    ./scripts/basic_upload_apks_service_account.py ch.opengis.qfield ${ASSETS}
+    ./scripts/basic_upload_apks_service_account.py ch.opengis.qfield-dev ${ASSETS}
```
Comment (3nids) [inline scripts/deploy.sh]: @m-kuhn @mbernasocchi is this the channel we can push to?

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-781-9e06b836ad5d0a78d95abbb40de9986bfd09b6c0-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-781-9e06b836ad5d0a78d95abbb40de9986bfd09b6c0-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-781-9e06b836ad5d0a78d95abbb40de9986bfd09b6c0-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-781-9e06b836ad5d0a78d95abbb40de9986bfd09b6c0-x86_64.apk)

File: scripts/deploy.sh (hunk 13 above)
Comment (m-kuhn) [reply scripts/deploy.sh]: sounds good to me :+1:

Comment (3nids) [reply scripts/deploy.sh]: do you want to wait for Marco's review or we merge and see what happens?

Comment (3nids) [reply scripts/deploy.sh]: (still need an approval before being able to merge)
